library/Vanilla/Models/ModelCache.php

```diff
@@ -0,0 +1,170 @@
+
+    /** @var string */
+    private $cacheNameSpace;
+
+    /** @var number */
+    private $defaultCacheOptions;
+
+    /** @var int */
+    private $incrementingKey;
+
+    /** @var \Gdn_Cache */
+    private $cache;
+
+    /** @var bool */
+    private $isFeatureDisabled;
+
+    /**
+     * Constructor.
+     *
+     * @param string $cacheNameSpace Namespace to use in the cache.
+     * @param \Gdn_Cache $cache The cache instance.
+     * @param array|null $defaultCacheOptions Default options to apply for storing cache values.
+     */
+    public function __construct(string $cacheNameSpace, \Gdn_Cache $cache, ?array $defaultCacheOptions = []) {
+        $this->cache = $cache;
+        $this->cacheNameSpace = $cacheNameSpace;
+        $this->defaultCacheOptions = array_merge(self::GLOBAL_DEFAULT_OPTIONS, $defaultCacheOptions ?? []);
+        $this->isFeatureDisabled = FeatureFlagHelper::featureEnabled(self::DISABLE_FEATURE_FLAG);
+    }
+
+    /**
+     * Create a cache key for some parameters.
+     *
+     * @param array $keyArgs Some arguments to generate the cache key from.
+     *
+     * @return string
+     */
+    public function createCacheKey(array $keyArgs): string {
+        $key = $this->cacheNameSpace . '-' . $this->getIncrementingKey() . '-' . sha1(json_encode($keyArgs));
+        return $key;
+    }
+
+    /**
+     * Try to get a cached record.
+     *
+     * If the record can't be found, we hydrate it with the $hydrate callable and return it.
+     *
+     * @param array $keyArgs The arguments to build the cache key.
+     * @param callable $hydrate A callable to hydrate the cache.
+     * @param array|null $cacheOptions Options for the cache storage.
+     *
+     * @return mixed
+     */
+    public function getCachedOrHydrate(array $keyArgs, callable $hydrate, ?array $cacheOptions = []) {
+        if ($this->isFeatureDisabled) {
+            return $hydrate();
+        }
+
+        $key = $this->createCacheKey($keyArgs);
+        $result = $this->cache->get($key);
+
+        if ($result === \Gdn_Cache::CACHEOP_FAILURE) {
+            $result = $hydrate();
+            $options = array_merge($this->defaultCacheOptions, $cacheOptions);
+            $this->cache->store($key, serialize($result), $options);
+        } else {
+            $result = unserialize($result);
+        }
+
+        return $result;
+    }
+
+    /**
+     * Invalidate all cached results for this cache.
+     */
+    public function invalidateAll() {
+        $this->rolloverIncrementingKey();
+    }
+
+    /**
+     * Create a pipeline processor for invalidating the entire cache on every record.
+     *
+     * @return ModelCacheInvalidationProcessor
+     */
+    public function createInvalidationProcessor(): ModelCacheInvalidationProcessor {
+        return new ModelCacheInvalidationProcessor($this);
+    }
+
+    /**
+     * @param \Gdn_Cache $cache
+     */
+    public function setCache(\Gdn_Cache $cache): void {
```

Comment:
Call this in the constructor and you get your lines covered ;)

Reply:
Done.